src/Plutarch/Test/QuickCheck/Modifiers.hs

```diff
@@ -292,6 +307,160 @@ instance Function (GenValue adaMod otherMod) where
+ @since 2.1.4
+-}
+deriving via (Sum Integer) instance Semigroup (TimeDelta NonPositive n)
+
+{- | Non-negative deltas are monoids with the zero delta as the identity.
+
+ @since 2.1.4
+-}
+deriving via (Sum Integer) instance Monoid (TimeDelta NonNegative n)
+
+{- | Non-positive deltas are monoids with the zero delta as the identity.
+
+ @since 2.1.4
+-}
+deriving via (Sum Integer) instance Monoid (TimeDelta NonPositive n)
+
+-- | @since 2.1.4
+instance (KnownNat n, 1 <= n) => Arbitrary (TimeDelta Positive n) where
+  {-# INLINEABLE arbitrary #-}
+  arbitrary = coerce <$> Gen.chooseInteger (1, integerVal @n)
+  {-# INLINEABLE shrink #-}
+  shrink = coerce @(Positive Integer -> [Positive Integer]) shrink
+
+-- | @since 2.1.4
+instance (KnownNat n, 1 <= n) => Arbitrary (TimeDelta Negative n) where
+  {-# INLINEABLE arbitrary #-}
+  arbitrary = coerce <$> Gen.chooseInteger (negate (integerVal @n), negate 1)
+  {-# INLINEABLE shrink #-}
+  shrink = coerce @(Negative Integer -> [Negative Integer]) shrink
+
+-- | @since 2.1.4
+instance (KnownNat n) => Arbitrary (TimeDelta NonPositive n) where
+  {-# INLINEABLE arbitrary #-}
+  arbitrary = coerce <$> Gen.chooseInteger (negate (integerVal @n), 0)
+  {-# INLINEABLE shrink #-}
+  shrink = coerce @(NonPositive Integer -> [NonPositive Integer]) shrink
+
+-- | @since 2.1.4
+instance (KnownNat n) => Arbitrary (TimeDelta NonNegative n) where
+  {-# INLINEABLE arbitrary #-}
+  arbitrary = coerce <$> Gen.chooseInteger (0, integerVal @n)
+  {-# INLINEABLE shrink #-}
+  shrink = coerce @(NonNegative Integer -> [NonNegative Integer]) shrink
+
+-- | @since 2.1.4
+deriving via
+  (mod Integer)
+  instance
+    ( forall (a :: Type) (b :: Type).
+      ( Coercible a b => Coercible (mod a) (mod b)
+      )
+    , CoArbitrary (mod Integer)
+    ) =>
+    CoArbitrary (TimeDelta mod n)
+
+-- | @since 2.1.4
+instance
+  (forall (a :: Type) (b :: Type). Coercible a b => Coercible (mod a) b) =>
+  Function (TimeDelta mod n)
+  where
+  {-# INLINEABLE function #-}
+  function = functionMap (coerce @(TimeDelta mod n) @Integer) coerce
+
+{- | A CPS-style 'Testable' handler for applying a 'TimeDelta' to a 'POSIXTime'.
+ If the application of the delta to the time would produce an impossible
+ result (that is, the resulting time is negative), this will automatically
+ fail the property test with an informative message; otherwise, it will apply
+ the delta, and produce a new 'POSIXTime', which it will pass to the function
+ argument to determine the outcome.
+
+ @since 2.1.4
+-}
+withTimeDelta ::
+  forall (n :: Nat) (mod :: Type -> Type).
+  (Coercible (mod POSIXTime) Integer) =>
+  TimeDelta mod n ->
+  POSIXTime ->
+  (POSIXTime -> Property) ->
+  Property
+withTimeDelta (TimeDelta d) time f = case signum (time + coerce d) of
+  (-1) -> counterexample badConversion . property $ False
+  d' -> f . coerce $ d'
+  where
+    badConversion :: String
+    badConversion =
+      "Applying a TimeDelta would yield an impossible time.\n"
```

Comment:
Since negative time is only instance of impossibility, can we say "Applying a TimeDelta would yield a negative time".

Reply:
Yeah, good call.